server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,874 @@
+                if (latestRevision != null) {
+                    future.complete(latestRevision);
+                } else {
+                    commitWatchers.add(normLastKnownRevision, pathPattern, future);
+                }
+            } finally {
+                readUnlock();
+            }
+        }, repositoryWorker).exceptionally(cause -> {
+            future.completeExceptionally(cause);
+            return null;
+        });
+
+        return future;
+    }
+
+    private void readLock() {
+        rwLock.readLock().lock();
+        if (closePending.get() != null) {
+            rwLock.readLock().unlock();
+            throw closePending.get().get();
+        }
+    }
+
+    private void readUnlock() {
+        rwLock.readLock().unlock();
+    }
+
+    private void writeLock() {
+        rwLock.writeLock().lock();
+        if (closePending.get() != null) {
+            writeUnLock();
+            throw closePending.get().get();
+        }
+    }
+
+    private void writeUnLock() {
+        rwLock.writeLock().unlock();
+    }
+
+    private static Commit currentInitialCommit(InternalRepository primaryRepo) {
+        final Revision firstRevision = primaryRepo.firstRevision();
+        final RevisionRange range = new RevisionRange(firstRevision, firstRevision);
+        return primaryRepo.listCommits(ALL_PATH, 1, range).get(0);
+    }
+
+    @Override
+    public Revision shouldCreateRollingRepository(int minRetentionCommits, int minRetentionDays) {
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        return shouldCreateRollingRepository(repo.headRevision(), minRetentionCommits, minRetentionDays);
+    }
+
+    @Nullable
+    private Revision shouldCreateRollingRepository(Revision headRevision, int minRetentionCommits,
+                                                   int minRetentionDays) {
+        // TODO(minwoox): provide a way to set different minRetentionCommits and minRetentionDays
+        //                in each repository.
+        if ((minRetentionCommits == 0 && minRetentionDays == 0) ||
+            (minRetentionCommits == Integer.MAX_VALUE && minRetentionDays == Integer.MAX_VALUE)) {
+            // Not enabled.
+            return null;
+        }
+
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        if (exceedsMinRetention(repo, headRevision, minRetentionCommits, minRetentionDays)) {
+            return headRevision;
+        }
+        return null;
+    }
+
+    private static boolean exceedsMinRetention(InternalRepository repo, Revision headRevision,
+                                               int minRetentionCommits, int minRetentionDays) {
+        final Revision firstRevision = repo.firstRevision();
+        if (minRetentionCommits != 0) {
+            if (headRevision.major() - firstRevision.major() <= minRetentionCommits) {
+                return false;
```

Comment:
Q) My initial understanding was that both conditions (days and commits) have to be satisfied to start the commit retention process.
Don't we also have to look at `minRetentionDays` before deciding?